File: leetcode/merge-two-sorted/merge.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next


class Solution:
    def sort(self, list: Optional[ListNode]) -> Optional[ListNode]:
        head = list
        curr = list.next

        while curr is not None:
            key = curr.val
            prev = head

            while curr != prev and prev.val > key:
                # ------ valor anterior maior que atual, troca -------
                # salva uma copia do anterior
                tmp = prev.val
                # valor anterior agora é o atual
                prev.val = key
                # o novo valor agora é a copia salva
                key = tmp
                # o proximo agora recebe o ponteiro do anterior
                curr = prev.next

            # se prev <= key, a validacao continua e pula para o proximo
            curr.val = key
            curr = curr.next

        return head

    def bubbleSort(self, list: Optional[ListNode]) -> Optional[ListNode]:
        # coming soon
        return list

    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        # verificar se as listas existem
        if not list1:
            return list2
        if not list2:
            return list1

        # definir um head para nova lista (valor inicial da lista ligada)
        if list1.val <= list2.val:
            head = list1
            list1 = list1.next

        if list2.val <= list2.val:
            head = list2
            list2 = list2.next

        current = head

        while list1 and list2:
            # elemento 1 é menor que o 2?
            if list1.val <= list2.val:
                # se sim, ele é o novo atual (uma vez que a lista deve estar ordenada)
                current.next = list1
                # pula para o proximo da lista 1
                list1 = list1.next
            else:
                # caso contrario, o proixmo item é o elemento da lista 2
                current.next = list2
                # pula para verificar o proximo elemento da lista 2
                list2 = list2.next

            # pular para o próximo
            current = current.next

        # retorna o penulitimo valor da lista, dado que trabalhamos com prev/curr
        current.next = list1 if list1 else list2

        return head


a = ListNode(1)
a.next = ListNode(3)
a.next.next = ListNode(5)

b = ListNode(0)
b.next = ListNode(2)
b.next.next = ListNode(4)
b.next.next.next = ListNode(6)

s = Solution()
items = s.mergeTwoLists(a, b)
# items = s.sort(b)

while items is not None:
    if items.val is not None:
        print(items.val)
    items = items.next

File: leetcode/merge-two-sorted/merge.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode
{
    int val;
    struct ListNode *next;
};

struct ListNode *mergeTwoLists(struct ListNode *list1, struct ListNode *list2)
{
    struct ListNode *head = NULL;
    struct ListNode *current = NULL;

    if (list1 == NULL) {
        return list2;
    }

    if (list2 == NULL) {
        return list1;
    }

    if (list1->val <= list2->val) {
        head = list1;
        list1 = list1->next;
    } else {
        head = list2;
        list2 = list2->next;
    }

    current = head;

    while (list1 != NULL && list2 != NULL) {
        if (list1->val <= list2->val) {
            current->next = list1;
            list1 = list1->next;
        } else {
            current->next = list2;
            list2 = list2->next;
        }

        current = current->next;
    }
    
    if (list1 != NULL) current->next = list1;
    if (list2 != NULL) current->next = list2;
    
    return head;
}

void printList(struct ListNode* head) {
    while (head != NULL) {
        printf("%d -> ", head->val);
        head = head->next;
    }
    printf("NULL\n");
}

struct ListNode* createNode(int val) {
    struct ListNode* node = (struct ListNode*)malloc(sizeof(struct ListNode));
    node->val = val;
    node->next = NULL;
    return node;
}

int main() {
    struct ListNode* l1 = createNode(1);
    l1->next = createNode(2);
    l1->next->next = createNode(4);

    struct ListNode* l2 = createNode(1);
    l2->next = createNode(3);
    l2->next->next = createNode(4);

    printf("List 1: ");
    printList(l1);

    printf("List 2: ");
    printList(l2);

    struct ListNode* merged = mergeTwoLists(l1, l2);

    printf("Merged: ");
    printList(merged);

    return 0;
}